src/storage/src/source/sql_server/replication.rs

```diff
@@ -125,14 +135,44 @@ pub(crate) fn render<G: Scope<Timestamp = Lsn>>(
 
             // Snapshot any instances that require it.
             let snapshot_lsn = {
-                tracing::info!(?needs_snapshot, "starting snapshot");
-                let (snapshot_lsn, snapshot_streams) =
+                // Small helper closure.
+                let emit_stats = |cap, known: usize, total: usize| {
+                    let update = ProgressStatisticsUpdate::Snapshot {
+                        records_known: u64::cast_from(known),
+                        records_staged: u64::cast_from(total),
+                    };
+                    tracing::debug!(?config.id, %known, %total, "snapshot progress");
+                    stats_output.give(cap, update);
+                };
+
+                tracing::debug!(?config.id, ?needs_snapshot, "starting snapshot");
+                // Eagerly emit an event if we have tables to snapshot.
+                if !needs_snapshot.is_empty() {
+                    emit_stats(&stats_cap[0], 0, 0);
+                }
+
+                let (snapshot_lsn, snapshot_stats, snapshot_streams) =
                     cdc_handle.snapshot(Some(needs_snapshot)).await?;
                 let snapshot_cap = data_cap_set.delayed(&snapshot_lsn);
 
+                // As we stream rows for the snapshot we'll track the total we've seen.
+                let mut records_total: usize = 0;
+                let records_known = snapshot_stats.values().sum();
+                let update_granularity =
+                    SNAPSHOT_STATS_GRANULARITY.handle(config.config.config_set());
+                if !snapshot_stats.is_empty() {
+                    emit_stats(&stats_cap[0], records_known, 0);
+                }
+
+                // Begin streaming our snapshots!
                 let mut snapshot_streams = std::pin::pin!(snapshot_streams);
                 while let Some((capture_instance, data)) = snapshot_streams.next().await {
                     let sql_server_row = data.map_err(TransientError::from)?;
+                    records_total = records_total.saturating_add(1);
+
+                    if records_total % update_granularity.get() == 0 {
```

Comment:
This will behave unpredictably depending on machine size, network speed, compression, row size, etc. I suggest we report things based on an `Instant::now()` value where we check whether some threshold time has elapsed

Reply:
Totally agree that this could change depending on machine size, network speed, etc. but I'm not totally sure what your proposal is? Are you suggesting every X seconds (based on `Instant::now()` timings) we report how many updates we've ingested thus far?

Also FWIW I went with this implementation because it appears to be what both [Postgres](https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/postgres/snapshot.rs#L489-L499) and [MySQL](https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/mysql/snapshot.rs#L447-L457) do too. Although I could be missing something here!

Reply:
Oh interesting. I didn't know we do this in pg and mysql and in a non-configurable way, yikes! Yes I was envisioning something like:

```rust
let mut last_report = Instant::now();
while .... {
   if last_report.elapsed() > SNAPSHOT_PROGRESS_REPORT_INTERVAL.get() {
       last_report = Instant::now();
       // reporting logic
   }
   ...
}
```

Reply:
Ahhh I see where you're coming from now. Is your concern that we'll emit events different rates, possibly too fast? Not that the events themselves have different values? At first I was thinking about the later and was confused how a time based approach would help, but the former is definitely better for a constant rate of emitting. Updating now!